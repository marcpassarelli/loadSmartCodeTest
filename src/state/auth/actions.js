import { Platform } from 'react-native';
import { authorize } from 'react-native-app-auth';
import { configAndroid, configIos } from '../../config';
import { getUser } from '../user/actions';

export const loginSuccess = (authInfo) => {
  return {
    type: 'LOGIN_SUCCESS',
    authInfo,
  };
};

export const loginFailed = (error) => {
  return {
    type: 'LOGIN_FAILED',
    error,
  };
};

export const logout = () => {
  return {
    type: 'LOGOUT',
  };
};

export const loginGithub = () => {
  return async (dispatch) => {
    try {
      const authInfo =
        Platform.OS === 'ios'
          ? await authorize(configIos)
          : await authorize(configAndroid);
      dispatch(getUser(authInfo.accessToken));
      dispatch(loginSuccess(authInfo));
    } catch (error) {
      dispatch(loginFailed(error));
    }
  };
};
