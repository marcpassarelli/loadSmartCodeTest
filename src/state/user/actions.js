export const setUser = (userInfo) => {
  return {
    type: 'SET_USER',
    userInfo,
  };
};

export const getUser = (accessToken) => {
  return (dispatch) => {
    return fetch('https://api.github.com/user', {
      method: 'GET',
      headers: {
        Authorization: 'token ' + accessToken,
      },
    })
      .then((response) => response.json())
      .then((userInfo) => {
        dispatch(setUser(userInfo));
      })
      .catch((error) => {
        console.log(error);
      });
  };
};
