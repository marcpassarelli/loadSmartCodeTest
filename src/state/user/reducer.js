const initialState = {
  userInfo: {},
};

const userReducer = (state = initialState, action) => {
  switch (action.type) {
    case 'SET_USER':
      return {
        ...state,
        userInfo: action.userInfo,
      };
      break;
    default:
      return state;
  }
};

export default userReducer;
