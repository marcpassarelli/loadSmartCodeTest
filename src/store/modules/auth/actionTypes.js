const actionTypes = {
  LOGIN_SUCCESS: 'LOGIN_SUCCESS',
  LOGIN_FAILED: 'LOGIN_FAILED',
  LOGOUT: 'LOGOUT',
};

export default actionTypes;
