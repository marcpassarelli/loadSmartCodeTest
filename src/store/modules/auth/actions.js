import { Platform } from 'react-native';
import { authorize, refresh } from 'react-native-app-auth';
import { configAndroid, configIos } from '../../../config';
import actionTypes from './actionTypes';

export const loginSuccess = (authInfo) => ({
  type: actionTypes.LOGIN_SUCCESS,
  authInfo,
});

export const loginFailed = (error) => ({
  type: actionTypes.LOGIN_FAILED,
  error,
});

export const logout = () => ({
  type: actionTypes.LOGOUT,
});

export const getNewAccessToken = (refreshToken) => {
  return async (dispatch) => {
    try {
      const config = Platform.OS === 'ios' ? configIos : configAndroid;
      const authInfo = !refreshToken
        ? await authorize(config)
        : await refresh(config, refreshToken);

      dispatch(loginSuccess(authInfo));
    } catch (error) {
      dispatch(loginFailed(error));
    }
  };
};
