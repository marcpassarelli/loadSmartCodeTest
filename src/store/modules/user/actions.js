import actionTypes from './actionTypes';

export const setUser = (userInfo) => {
  return {
    type: actionTypes.SET_USER,
    userInfo,
  };
};

export const getUser = (accessToken) => (dispatch) =>
  fetch('https://api.github.com/user', {
    method: 'GET',
    headers: {
      Authorization: 'token ' + accessToken,
    },
  })
    .then((response) => response.json())
    .then((userInfo) => {
      dispatch(setUser(userInfo));
    })
    .catch((error) => {
      console.log(error);
    });
