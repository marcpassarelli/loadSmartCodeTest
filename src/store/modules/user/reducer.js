import actionTypes from './actionTypes';

export const initialState = {
  userInfo: {},
};

const userReducer = (state = initialState, action) => {
  switch (action.type) {
    case actionTypes.SET_USER:
      return {
        ...state,
        userInfo: action.userInfo,
      };
    default:
      return state;
  }
};

export default userReducer;
