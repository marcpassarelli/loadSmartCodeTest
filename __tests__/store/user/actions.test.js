import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock from 'fetch-mock-jest';
import expect from 'expect'; // You can use any testing library
import actionTypes from '../../../src/store/modules/user/actionTypes';
import { getUser } from '../../../src/store/modules/user/actions';
import userReducer from '../../../src/store/modules/user/reducer';

const middleware = [thunk];
const mockStore = configureMockStore(middleware);

describe('async actions', () => {
  afterEach(() => {
    fetchMock.restore();
  });

  it('creates SET_USER when fetching for users info has been done', () => {
    fetchMock.getOnce('https://api.github.com/user', {
      body: {
        name: 'Marc Victor',
        avatar_url: 'photo.jpg',
        login: 'marcpassarelli',
        location: 'SP',
        bio: 'Hello',
      },
    });

    const expectedActions = [
      {
        type: actionTypes.SET_USER,
        userInfo: {
          name: 'Marc Victor',
          avatar_url: 'photo.jpg',
          login: 'marcpassarelli',
          location: 'SP',
          bio: 'Hello',
        },
      },
    ];
    const store = mockStore({});

    return store.dispatch(getUser()).then(() => {
      // return of async actions
      expect(store.getActions()).toEqual(expectedActions);
      expect(store.getActions()).toMatchSnapshot();
    });
  });
});
