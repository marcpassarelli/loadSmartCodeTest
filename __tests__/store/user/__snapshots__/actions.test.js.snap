// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`async actions creates SET_USER when fetching for users info has been done 1`] = `
Array [
  Object {
    "type": "SET_USER",
    "userInfo": Object {
      "avatar_url": "photo.jpg",
      "bio": "Hello",
      "location": "SP",
      "login": "marcpassarelli",
      "name": "Marc Victor",
    },
  },
]
`;
